PythonVisualizations/SortingBase.py

```diff
@@ -0,0 +1,420 @@
+        cell_center = self.cellCenter(index)
+        level_spacing = self.VARIABLE_FONT[1]
+        x = cell_center[0]
+        if level > 0:
+            y0 = cell_coords[1] - self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[1] - self.CELL_SIZE * 3 // 10
+        else:
+            y0 = cell_coords[3] + self.CELL_SIZE * 3 // 5 - level * level_spacing
+            y1 = cell_coords[3] + self.CELL_SIZE * 3 // 10
+        arrow = self.canvas.create_line(
+            x, y0, x, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x + 2, y0, text=name, anchor=SW if level > 0 else NW,
+                font=self.VARIABLE_FONT, fill=color)
+        return (arrow, label) if name else (arrow,)  
+ 
+
+    # ARRAY FUNCTIONALITY
+    def insert(self, val):
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+        
+        # Inserting when there is no room on the canvas
+        if self.window.winfo_width() <= self.ARRAY_X0 + (
+            (len(self.list)+1) * self.CELL_SIZE):
+            # change cell width to 20
+            self.CELL_WIDTH = 20
+            # if this is the first insert that will change the width 
+            if not self.changeSize:
+                self.changeSize = True 
+                # redraw the existing cells to be narrower
+                self.fixCells()
+                self.redrawArrayCells()
+                # delete the cell values 
+                self.canvas.delete("cellVal")
+            
+        # If array needs to grow, add cells:
+        while self.size < len(self.list) + 1:
+            self.size += 1
+            self.createArrayCell(len(self.list))
+
+        # draw an index pointing to the last cell
+        indexDisplay = self.createIndex(len(self.list), 'nItems')
+        callEnviron |= set(indexDisplay)
+
+        # create new cell and cell value display objects
+        toPositions = (self.fillCoords(val, self.cellCoords(len(self.list))),
+                       self.cellCenter(len(self.list)))
+
+        # Animate arrival of new value from operations panel area
+        canvasDimensions = self.widgetDimensions(self.canvas)
+        startPosition = add_vector(
+            [canvasDimensions[0] // 2 - self.CELL_WIDTH, canvasDimensions[1]] * 2,
+            (0, 0) + (self.CELL_WIDTH - self.CELL_BORDER, 2.5*self.CELL_SIZE - self.CELL_BORDER))
+        
+        cellPair = self.createCellValue(startPosition, val)
+        callEnviron |= set(cellPair)
+        # if it is the smaller size cell
+        if self.changeSize:
+            # remove the generated cell value 
+            self.canvas.delete(cellPair[1])
+            callEnviron.remove(cellPair[1])
+            self.moveItemsTo((cellPair[0],), toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
+        else:
+            self.moveItemsTo(cellPair, toPositions, steps=self.CELL_SIZE, sleepTime=0.01)
```

Comment:
This is kind of awkward code to test whether the array cells are smaller and delete the text part of the cell pair.  I think it would be nicer if the `cellPair` created by `createCellValue` took care of whether to create the text item for the key value in that cell based on the cell width.  It could return a tuple like `(`_<canvas_rectangle>_`, None)` for cell pairs that don't have enough space for text.  That would mean routines like `cleanUp()`, `copyCanvasItem()`, and `moveItems`*`()` would have to test if the items passed to them are `None` before trying to do operations, but its conceptually easier to understand the code.  Having `None` for the text part of the cell pair would mean lines 166-171 could be removed.  (Of course, that's a lot of changes to make on other parts of the code, but I think it's worth being able to have the text be optional).

On a separate topic, it's not a good idea to duplicate the code that calculates the size and shape of the `startPosition` like on lines 159-161.   That means that any future changes to how cells are sized and shaped has to be repeated everywhere that calculation occurs.  It would be better to make use of the size and shape calculated for the `toPositions` and shift them off the canvas.  For example:
```
startPosition = add_vector(toPositions[1],
            subtract_vector([canvasDimensions[0] // 2 - self.CELL_WIDTH, canvasDimensions[1]] * 2,
                            toPositions[1]))
```

Reply:
Fixed in commit 8169bd2 and earlier